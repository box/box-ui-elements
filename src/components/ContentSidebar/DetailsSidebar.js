/**
 * @flow
 * @file Details sidebar component
 * @author Box
 */

import React from 'react';
import getProp from 'lodash/get';
import { FormattedMessage, injectIntl } from 'react-intl';
import ItemProperties from 'box-react-ui/lib/features/item-details/ItemProperties';
import getFileSize from 'box-react-ui/lib/utils/getFileSize';
import messages from '../messages';
import SidebarSection from './SidebarSection';
import SidebarContent from './SidebarContent';
import SidebarSkills from './Skills/SidebarSkills';
import SidebarNotices from './SidebarNotices';
import type { BoxItem } from '../../flowTypes';
import './DetailsSidebar.scss';

type Props = {
    file: BoxItem,
    getPreviewer: Function,
    hasTitle: boolean,
    hasSkills: boolean,
    hasNotices: boolean,
    hasProperties: boolean,
    hasMetadata: boolean,
    hasAccessStats: boolean,
    hasClassification: boolean,
    rootElement: HTMLElement,
    appElement: HTMLElement,
    onInteraction: Function,
    onDescriptionChange: Function,
    intl: any
};

/* eslint-disable jsx-a11y/label-has-for */
const DetailsSidebar = ({
    file,
    getPreviewer,
    hasTitle,
    hasSkills,
    hasProperties,
    hasNotices,
    hasMetadata,
    hasAccessStats,
    hasClassification,
    rootElement,
    appElement,
    onInteraction,
    onDescriptionChange,
    intl
}: Props) => {
    if (!hasSkills && !hasProperties && !hasMetadata && !hasAccessStats && !hasClassification && !hasNotices) {
        return null;
    }

    const onDescriptionChangeEditable = getProp(file, 'permissions.can_rename') ? onDescriptionChange : undefined;

    return (
        <SidebarContent hasTitle={hasTitle} title={<FormattedMessage {...messages.sidebarDetailsTitle} />}>
            {hasNotices && <SidebarNotices file={file} />}
            {hasSkills && (
                <SidebarSkills
                    metadata={file.metadata}
                    getPreviewer={getPreviewer}
                    rootElement={rootElement}
                    appElement={appElement}
                    onInteraction={onInteraction}
                />
            )}
            {hasProperties && (
                <SidebarSection title={<FormattedMessage {...messages.sidebarProperties} />}>
                    <ItemProperties
                        createdAt={file.created_at}
                        description={file.description}
                        modifiedAt={file.modified_at}
                        owner={getProp(file, 'owned_by.name')}
                        size={getFileSize(file.size, intl.locale)}
                        uploader={getProp(file, 'created_by.name')}
                        onDescriptionChange={onDescriptionChangeEditable}
                        descriptionTextareaProps={{ maxLength: '255' }}
                    />
                </SidebarSection>
            )}
        </SidebarContent>
    );
};

export default injectIntl(DetailsSidebar);
