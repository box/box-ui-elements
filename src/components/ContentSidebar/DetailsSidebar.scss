@import '../variables';
@import '~box-react-ui/lib/styles/common/links';

.be .item-properties dd {
    margin-left: 0;

    &:last-child {
        margin-bottom: 0;
    }
}

.bcs-details-content {
    margin: 0 10px 20px 20px;
}

.bcs-details-content:empty {
    margin: 0;
}
