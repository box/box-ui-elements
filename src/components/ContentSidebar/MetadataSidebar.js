/**
 * @flow
 * @file Metadata sidebar component
 * @author Box
 */

import * as React from 'react';
import { FormattedMessage } from 'react-intl';
import MetadataInstanceEditor from 'box-react-ui/lib/features/metadata-instance-editor/MetadataInstanceEditor';
import TemplateDropdown from 'box-react-ui/lib/features/metadata-instance-editor/TemplateDropdown';
import LoadingIndicator from 'box-react-ui/lib/components/loading-indicator/LoadingIndicator';
import LoadingIndicatorWrapper from 'box-react-ui/lib/components/loading-indicator/LoadingIndicatorWrapper';
import InlineError from 'box-react-ui/lib/components/inline-error/InlineError';
import messages from '../messages';
import SidebarContent from './SidebarContent';
import APIContext from '../APIContext';
import API from '../../api';
import './MetadataSidebar.scss';

type ExternalProps = {
    getMetadata?: Function
};

type PropsWithoutContext = {
    file: BoxItem
} & ExternalProps;

type Props = {
    api: API
} & PropsWithoutContext;

type State = {
    editors?: Array<MetadataEditor>,
    templates?: Array<MetadataEditorTemplate>,
    isLoading: boolean,
    hasError: boolean
};

class MetadataSidebar extends React.PureComponent<Props, State> {
    state = {
        isLoading: false,
        hasError: false
    };

    componentDidMount() {
        this.getMetadataEditors();
    }

    /**
     * Sets the error state to true
     *
     * @return {void}
     */
    errorCallback = (): void => {
        this.setState({ isLoading: false, hasError: true });
    };

    /**
     * Fetches the metadata editors
     *
     * @return {void}
     */
    getMetadataEditors = (): void => {
        const { api, file, getMetadata }: Props = this.props;
        api.getMetadataAPI(true).getEditors(
            file,
            ({ editors, templates }: { editors: Array<MetadataEditor>, templates: Array<MetadataEditorTemplate> }) => {
                this.setState({ templates, editors: editors.slice(0), isLoading: false, hasError: false });
            },
            this.errorCallback,
            getMetadata
        );
    };

    /**
     * Checks upload permission
     *
     * @return {boolean} - true if metadata can be edited
     */
    canEdit(): boolean {
        const { file }: Props = this.props;
        const { permissions = {} }: BoxItem = file;
        const { can_upload }: BoxItemPermission = permissions;
        return !!can_upload;
    }

    /**
     * Instance remove handler
     *
     * @param {number} id - instance id
     * @return {void}
     */
    onRemove = (id: string): void => {
        const { api, file }: Props = this.props;
        const { editors = [] }: State = this.state;
        const editor = editors.find(({ instance }) => instance.id === id);
        if (!editor) {
            return;
        }

        api.getMetadataAPI(false).deleteMetadata(file, editor.template, this.getMetadataEditors, this.errorCallback);
    };

    /**
     * Instance add handler
     *
     * @param {Object} template - instance template
     * @return {void}
     */
    onAdd = (template: MetadataEditorTemplate) => {
        const { api, file }: Props = this.props;
        this.setState({ isLoading: true });
        api.getMetadataAPI(false).createMetadata(file, template, this.getMetadataEditors, this.errorCallback);
    };

    /**
     * Instance save handler
     *
     * @param {number} id - instance id
     * @param {Array} ops - json patch ops
     * @return {void}
     */
    onSave = (id: string, ops: JsonPatchData): void => {
        const { api, file }: Props = this.props;
        const { editors = [] }: State = this.state;
        const editor = editors.find(({ instance }) => instance.id === id);
        if (!editor) {
            return;
        }

        api
            .getMetadataAPI(false)
            .updateMetadata(file, editor.template, ops, this.getMetadataEditors, this.errorCallback);
    };

    /**
     * Instance dirty handler
     *
     * @param {number} id - instance id
     * @param {boolean} isDirty - instance dirty state
     * @return {void}
     */
    onModification = (id: string, isDirty: boolean) => {
        const { editors = [] }: State = this.state;
        const index = editors.findIndex(({ instance }) => instance.id === id);
        if (index === -1) {
            return;
        }

        const editor = { ...editors[index] };
        editor.isDirty = isDirty;
        const clone = editors.slice(0);
        clone.splice(index, 1, editor);
        this.setState({ editors: clone });
    };

    render() {
        const { editors, templates, isLoading, hasError }: State = this.state;
        const showEditor = !!templates && !!editors;
        const showLoadingIndicator = !hasError && !showEditor;
        const canEdit = this.canEdit();
        const showTemplateDropdown = showEditor && canEdit;

        return (
            <SidebarContent
                title={<FormattedMessage {...messages.sidebarMetadataTitle} />}
                actions={
                    showTemplateDropdown ? (
                        <TemplateDropdown
                            hasTemplates={templates && templates.length !== 0}
                            isDropdownBusy={false}
                            onAdd={this.onAdd}
                            templates={templates}
                            usedTemplates={editors && editors.map((editor) => editor.template)}
                        />
                    ) : null
                }
            >
                {hasError && (
                    <InlineError title={<FormattedMessage {...messages.sidebarMetadataErrorTitle} />}>
                        <FormattedMessage {...messages.sidebarMetadataErrorContent} />
                    </InlineError>
                )}
                {showLoadingIndicator && <LoadingIndicator />}
                {showEditor && (
                    <LoadingIndicatorWrapper isLoading={isLoading}>
                        <MetadataInstanceEditor
                            canAdd={canEdit}
                            editors={editors}
                            onSave={this.onSave}
                            onModification={this.onModification}
                            onAdd={this.onAdd}
                            onRemove={this.onRemove}
                            templates={templates}
                        />
                    </LoadingIndicatorWrapper>
                )}
            </SidebarContent>
        );
    }
}

export type MetadataSidebarProps = ExternalProps;
export { MetadataSidebar as MetadataSidebarComponent };
export default (props: PropsWithoutContext) => (
    <APIContext.Consumer>{(api) => <MetadataSidebar {...props} api={api} />}</APIContext.Consumer>
);
