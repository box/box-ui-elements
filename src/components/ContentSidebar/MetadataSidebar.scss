@import '../variables';

.be.bcs.bcs-metadata {
    .metadata-instance-editor {
        min-height: 400px;
        padding: 0 10px 20px 25px;
    }

    .metadata-instance-editor-header {
        display: none;
    }

    .metadata-instance-editor-field-read-only dd {
        margin: 0;
    }

    .metadata-instance-editor-instance-title-text {
        max-width: 210px;
        overflow: hidden;
        text-overflow: ellipsis;
        white-space: nowrap;
    }

    .collapsible-card.is-open .metadata-instance-editor-instance-title-text {
        max-width: 190px;
    }

    .inline-alert {
        margin: 0 10px 10px 25px;
    }
}
