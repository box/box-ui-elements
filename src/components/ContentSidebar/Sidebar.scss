@import '../variables';
@import '~box-react-ui/lib/styles/common/_inline-notifications';

.bcs {
    .tab-view {
        display: flex;
        flex: 1;
        flex-direction: column;
    }

    .tabs {
        margin: 0 30px;
    }

    .tab-panels {
        padding-top: 0;

        .tab-panel.is-selected {
            display: flex;
        }
    }
}
