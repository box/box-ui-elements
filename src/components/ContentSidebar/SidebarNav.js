/**
 * @flow
 * @file Preview sidebar nav component
 * @author Box
 */

import * as React from 'react';
import { FormattedMessage } from 'react-intl';
import IconMagicWand from 'box-react-ui/lib/icons/general/IconMagicWand';
import IconMetadataThick from 'box-react-ui/lib/icons/general/IconMetadataThick';
import IconDocInfo from 'box-react-ui/lib/icons/general/IconDocInfo';
import IconChatRound from 'box-react-ui/lib/icons/general/IconChatRound';
import SidebarNavButton from './SidebarNavButton';
import messages from '../messages';
import { SIDEBAR_NAV_TARGETS } from '../../interactionTargets';
import {
    SIDEBAR_VIEW_SKILLS,
    SIDEBAR_VIEW_ACTIVITY,
    SIDEBAR_VIEW_DETAILS,
    SIDEBAR_VIEW_METADATA,
} from '../../constants';
import './SidebarNav.scss';

type Props = {
    hasSkills: boolean,
    hasMetadata: boolean,
    hasActivityFeed: boolean,
    hasDetails: boolean,
    onToggle: Function,
    selectedView?: SidebarView,
};

const SidebarNav = ({
    hasSkills,
    hasMetadata,
    hasActivityFeed,
    hasDetails,
    onToggle,
    selectedView,
}: Props) => (
    <nav>
        {hasActivityFeed && (
            <SidebarNavButton
                tooltip={
                    <FormattedMessage {...messages.sidebarActivityTitle} />
                }
                onClick={() => onToggle(SIDEBAR_VIEW_ACTIVITY)}
                interactionTarget={SIDEBAR_NAV_TARGETS.ACTIVITY}
                isSelected={SIDEBAR_VIEW_ACTIVITY === selectedView}
            >
                <IconChatRound />
            </SidebarNavButton>
        )}
        {hasDetails && (
            <SidebarNavButton
                tooltip={<FormattedMessage {...messages.sidebarDetailsTitle} />}
                onClick={() => onToggle(SIDEBAR_VIEW_DETAILS)}
                interactionTarget={SIDEBAR_NAV_TARGETS.DETAILS}
                isSelected={SIDEBAR_VIEW_DETAILS === selectedView}
            >
                <IconDocInfo />
            </SidebarNavButton>
        )}
        {hasSkills && (
            <SidebarNavButton
                tooltip={<FormattedMessage {...messages.sidebarSkillsTitle} />}
                onClick={() => onToggle(SIDEBAR_VIEW_SKILLS)}
                interactionTarget={SIDEBAR_NAV_TARGETS.SKILLS}
                isSelected={SIDEBAR_VIEW_SKILLS === selectedView}
            >
                <IconMagicWand />
            </SidebarNavButton>
        )}
        {hasMetadata && (
            <SidebarNavButton
                tooltip={
                    <FormattedMessage {...messages.sidebarMetadataTitle} />
                }
                onClick={() => onToggle(SIDEBAR_VIEW_METADATA)}
                interactionTarget={SIDEBAR_NAV_TARGETS.METADATA}
                isSelected={SIDEBAR_VIEW_METADATA === selectedView}
            >
                <IconMetadataThick />
            </SidebarNavButton>
        )}
    </nav>
);

export default SidebarNav;
