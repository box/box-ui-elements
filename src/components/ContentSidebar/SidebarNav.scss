@import '../variables';

.bcs {
    nav {
        border-left: 1px solid $off-white;
        border-right: 1px solid $off-white;
    }
}
