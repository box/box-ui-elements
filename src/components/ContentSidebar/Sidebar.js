/**
 * @flow
 * @file Preview sidebar component
 * @author Box
 */

import React from 'react';
import { injectIntl } from 'react-intl';
import TabView from 'box-react-ui/lib/components/tab-view/TabView';
import Tab from 'box-react-ui/lib/components/tab-view/Tab';
import DetailsSidebar from './DetailsSidebar';
import { hasSkills as hasSkillsData } from './Skills/skillUtils';
import messages from '../messages';
import type { BoxItem } from '../../flowTypes';
import './Sidebar.scss';

type Props = {
    file: BoxItem,
    getPreviewer: Function,
    hasTitle: boolean,
    hasSkills: boolean,
    hasProperties: boolean,
    hasMetadata: boolean,
    hasNotices: boolean,
    hasAccessStats: boolean,
    hasClassification: boolean,
    hasActivityFeed: boolean,
    rootElement: HTMLElement,
    appElement: HTMLElement,
    onInteraction: Function,
    onDescriptionChange: Function,
    descriptionTextareaProps: Object,
    intl: any
};

const Sidebar = ({
    file,
    getPreviewer,
    hasTitle,
    hasSkills,
    hasProperties,
    hasMetadata,
    hasNotices,
    hasAccessStats,
    hasClassification,
    hasActivityFeed,
    rootElement,
    appElement,
    onInteraction,
    onDescriptionChange,
    intl
}: Props) => {
    const shouldShowSkills = hasSkills && hasSkillsData(file);

    const Details = (
        <DetailsSidebar
            file={file}
            getPreviewer={getPreviewer}
            hasTitle={hasTitle}
            hasSkills={shouldShowSkills}
            hasProperties={hasProperties}
            hasMetadata={hasMetadata}
            hasNotices={hasNotices}
            hasAccessStats={hasAccessStats}
            hasClassification={hasClassification}
            appElement={appElement}
            rootElement={rootElement}
            onInteraction={onInteraction}
            onDescriptionChange={onDescriptionChange}
        />
    );

    if (!hasActivityFeed) {
        return Details;
    }

    return (
        <TabView defaultSelectedIndex={shouldShowSkills ? 0 : 1}>
            <Tab title={intl.formatMessage(messages.sidebarDetailsTitle)}>{Details}</Tab>
            <Tab title='Activity' />
        </TabView>
    );
};

export default injectIntl(Sidebar);
