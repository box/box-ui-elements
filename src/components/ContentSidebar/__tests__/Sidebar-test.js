import React from 'react';
import { shallow } from 'enzyme';
import Sidebar from '../Sidebar';
import DetailsSidebar from '../DetailsSidebar';
import SkillsSidebar from '../SkillsSidebar';
import MetadataSidebar from '../MetadataSidebar';
import SidebarNav from '../SidebarNav';

jest.mock('../ActivitySidebar', () => 'ActivitySidebar');

describe('components/ContentSidebar/Skills/Sidebar', () => {
    const file = { id: 'id' };
    const getWrapper = props => shallow(<Sidebar file={file} {...props} />);

    test('should render no sidebar', () => {
        const wrapper = getWrapper();
        expect(wrapper.find(SidebarNav)).toHaveLength(1);
        expect(wrapper).toMatchSnapshot();
    });

    test('should render skills sidebar', () => {
        const wrapper = getWrapper({ hasSkills: true, view: 'skills' });
        expect(wrapper.find(SidebarNav)).toHaveLength(1);
        expect(wrapper.find(SkillsSidebar)).toHaveLength(1);
        expect(wrapper).toMatchSnapshot();
    });

    test('should render activity sidebar', () => {
        const wrapper = getWrapper({ hasActivityFeed: true, view: 'activity' });
        expect(wrapper.find(SidebarNav)).toHaveLength(1);
        expect(wrapper.find('ActivitySidebar')).toHaveLength(1);
        expect(wrapper).toMatchSnapshot();
    });

    test('should render details sidebar', () => {
        const wrapper = getWrapper({ hasDetails: true, view: 'details' });
        expect(wrapper.find(SidebarNav)).toHaveLength(1);
        expect(wrapper.find(DetailsSidebar)).toHaveLength(1);
        expect(wrapper).toMatchSnapshot();
    });

    test('should render metadata sidebar', () => {
        const wrapper = getWrapper({ hasMetadata: true, view: 'metadata' });
        expect(wrapper.find(SidebarNav)).toHaveLength(1);
        expect(wrapper.find(MetadataSidebar)).toHaveLength(1);
        expect(wrapper).toMatchSnapshot();
    });
});
