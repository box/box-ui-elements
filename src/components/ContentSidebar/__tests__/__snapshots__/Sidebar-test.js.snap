// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ContentSidebar/Skills/Sidebar should render activity sidebar 1`] = `
<React.Fragment>
  <SidebarNav
    hasActivityFeed={true}
    hasDetails={false}
    onToggle={[Function]}
    selectedView="activity"
  />
  <ActivitySidebar />
</React.Fragment>
`;

exports[`components/ContentSidebar/Skills/Sidebar should render details sidebar 1`] = `
<React.Fragment>
  <SidebarNav
    hasDetails={true}
    onToggle={[Function]}
    selectedView="details"
  />
  <DetailsSidebar
    hasProperties={true}
  />
</React.Fragment>
`;

exports[`components/ContentSidebar/Skills/Sidebar should render no sidebar 1`] = `
<React.Fragment>
  <SidebarNav
    hasDetails={false}
    onToggle={[Function]}
    selectedView="details"
  />
</React.Fragment>
`;

exports[`components/ContentSidebar/Skills/Sidebar should render skills sidebar 1`] = `
<React.Fragment>
  <SidebarNav
    hasDetails={false}
    hasSkills={true}
    onToggle={[Function]}
    selectedView="skills"
  />
  <SkillsSidebar />
</React.Fragment>
`;
